### DISCUSSION TITLE
How to handle CRDs that do not exist until HelmRelease is installed?

### DISCUSSION THREAD

**Discussion by ilya-git** [fluxcd/flux2#2282] [Q&A] [answered]
I have this trouble right now:

1. I have a gitops repository that has a HelmRelease `cert-manager`
2. I have a `ClusterIssuer` manifest (a CRD from `cert-manager`)
3. Kustomization fails with this error: "kustomize-controller  ClusterIssuer/letsencrypt-prod dry-run failed, error: no matches for kind "ClusterIssuer" in version "cert-manager.io/v1""

So dry-run fails because there is a CRD that it does not know about. The same CRD that will only be installed after kustomization is completed. It's kind of a dead loop here that I do not know how to resolve properly.

I can manually do some things like commenting the CRD, installing the helmrelease, etc, but I am wondering if there is a better solution to this "chicken-egg" problem

**Comment by ilya-git** [comment, marked as answer]
It seems the answer to the question lies here in the docs: https://fluxcd.io/docs/components/kustomize/kustomization/#kustomization-dependencies

**Reply by kingdonb** [reply]
That's exactly it, thank you 👍 

We had some discussion about what an ideal solution might look like here: https://github.com/fluxcd/helm-controller/issues/387

I was saying that because it's technically a feature request, it belongs in discussions instead. @jr0dd

But the canonical answer is, it is already a solved problem, you just have to use Kustomization `dependsOn` ordering. It is maybe less than ideal that you have to use two `Kustomizations` and one `HelmRelease` but it is the current state of the art in Flux, if you keep your CRDs in Helm.

**Comment by ilya-git** [comment]
Thanks for the detailed comment.

It is not very convenient though to solve it via dependencies though, since it is not possible (as far as I know at least) to have a kustomization with path to a helm release and then a customization that adds everything else (i.e. exclude). So that release would have to be in a special "pre-run" folder, while everything else would have to be in some "run" folder.

Or maybe I did misunderstand how it is supposed to work?

Perhaps some "run helm first" option can be applied as a quick fix? but there are probably other similar cases

**Reply by yebyen** [reply]
It's not very convenient you have that right, that is the subject of the linked thread 👍

https://github.com/fluxcd/helm-controller/issues/387

There's a natural enhancement that's impossible without some alternative to the existing Kustomization resource structure

**Comment by jceb** [comment]
I've also stumbled over this problem a number of times. What I find least optimal about multiple kustomizations is the difficulty they add to linting the whole cluster configuration with `datree` or `kubeconform`. Often enough, I make a mistake in a sub-kustomization, receive a postivie result from the pre-commit linter only to find out that the kustomization is actually broken.

My first attempt to solve the challenges of handling CRDs was to automate running `helm template` locally for helm charts and storing all CRDs and configurations directly in the repository. I'm so happy with that configuration that I'll go back to it as my main method of integrating third-party charts.
